compiler/rustc_mir_transform/src/simplify_static_switch.rs

```diff
@@ -0,0 +1,155 @@
+use super::MirPass;
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_index::IndexSlice;
+use rustc_middle::mir::traversal;
+use rustc_middle::mir::{
+    AggregateKind, BasicBlock, Body, Operand, Rvalue, StatementKind, TerminatorKind,
+};
+use rustc_middle::ty::TyCtxt;
+use rustc_session::Session;
+
+use super::simplify::remove_dead_blocks;
+
+/// This pass looks to optimize a pattern in MIR where a discriminant is
+/// constructed in one or more blocks with the same successor and then
+/// that discriminant is switched on in that successor block, in which case
+/// we can remove the need to switch on the discriminant at all because we
+/// statically know what target block will be taken for each constructed
+/// discriminant.
+///
+/// For example, the following CFG:
+/// ```text
+///     x = Foo::A(y); ---       Foo::A ---> ...
+///    /                  \            /
+/// ...                    --> switch x
+///    \                  /            \
+///     x = Foo::B(y); ---       Foo::B ---> ...
+/// ```
+/// would become:
+/// ```text
+///     x = Foo::A(y); --------- Foo::A ---> ...
+///    /
+/// ...
+///    \
+///     x = Foo::B(y); --------- Foo::B ---> ...
+/// ```
+pub struct SimplifyStaticSwitch;
+
+impl<'tcx> MirPass<'tcx> for SimplifyStaticSwitch {
+    fn name(&self) -> &'static str {
+        "SimplifyStaticSwitch"
+    }
+
+    fn is_enabled(&self, sess: &Session) -> bool {
+        sess.mir_opt_level() >= 2
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        debug!(body = %tcx.def_path_str(body.source.def_id()));
+
+        if simplify_static_switch(tcx, body) {
+            remove_dead_blocks(tcx, body);
+        }
+    }
+}
+
+#[instrument(level = "debug", skip(tcx, body), ret)]
+fn simplify_static_switch<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) -> bool {
+    let mut new_targets = FxHashMap::default();
+    let predecessors: &IndexSlice<_, _> = body.basic_blocks.predecessors();
+    for (block, data) in traversal::preorder(body) {
```

Comment:
Why this particular order?

Reply:
I thought it might be better than iterating normally but i have switched it back to just `iter_enumerated`.